src/Runtime/Distributions/Automata/Automaton.Builder.cs

```diff
@@ -0,0 +1,810 @@
+            /// </summary>
+            public StateBuilder this[int index] => new StateBuilder(this, index);
+
+            /// <summary>
+            /// Returns <see cref="StateBuilder"/> object for start state.
+            /// </summary>
+            public StateBuilder Start => this[this.StartStateIndex];
+
+            #endregion
+
+            # region Factory methods
+
+            /// <summary>
+            /// Factory method which creates <see cref="Builder"/> with single state.
+            /// </summary>
+            public static Builder Zero()
+            {
+                var builder = new Builder();
+                builder.SetToZero();
+                return builder;
+            }
+
+            /// <summary>
+            /// Factory method which creates <see cref="Builder"/> which contains all states and transitions
+            /// from given <paramref name="automaton"/>.
+            /// </summary>
+            public static Builder FromAutomaton(Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TThis> automaton)
+            {
+                var result = new Builder();
+                result.AddStates(automaton.States);
+                result.StartStateIndex = automaton.Start.Index;
+                return result;
+            }
+
+            /// <summary>
+            /// Factory method which creates <see cref="Builder"/> and inserts into it an automaton
+            /// matching given <paramref name="sequence"/> with some <paramref name="weight"/>.
+            /// </summary>
+            public static Builder ConstantOn(Weight weight, TSequence sequence)
+            {
+                var result = Builder.Zero();
+                result.Start.AddTransitionsForSequence(sequence).SetEndWeight(weight);
+                return result;
+            }
+
+            #endregion
+
+            #region States manipulation (AddState and RemoveState methods state family)
+
+            /// <summary>
+            /// Adds new state and returns <see cref="StateBuilder"/> for it.
+            /// </summary>
+            public StateBuilder AddState()
+            {
+                if (this.states.Count >= maxStateCount)
+                {
+                    throw new AutomatonTooLargeException(MaxStateCount);
+                }
+
+                var index = this.states.Count;
+                this.states.Add(
+                    new StateData
+                    {
+                        FirstTransition = -1,
+                        LastTransition = -1,
+                        EndWeight = Weight.Zero,
+                    });
+                return new StateBuilder(this, index);
+            }
+
+            /// <summary>
+            /// Adds <paramref name="count"/> new states/
+            /// </summary>
+            public void AddStates(int count)
+            {
+                for (var i = 0; i < count; ++i)
+                {
+                    AddState();
```

Comment:
I cannot remember the exact gc implications of not using the return object of a method in a loop, but I have a feeling you may want to avoid this.  Feel free to ignore this comment though...

Reply:
1. It actually returns a struct
2. It gets inlined and costs nothing, I measured